### PR TITLE
Initialise column types during appender creation

### PR DESCRIPTION
**Pull request by taniabogatsch**
As discussed in #168, this PR adds support for initializing the column types of a data chunk when creating an appender.

The PR builds on https://github.com/marcboeker/go-duckdb/pull/167. I'll undraft it once #167 is merged.

### CODE REVIEW

**Comment by taniabogatsch** [comment]
I added a map to track all as-of-now unsupported types in the appender. We can incrementally add these (if they make sense in Go).

File: appender.go [hunk 1]
```diff
@@ -562,7 +548,7 @@ func (a *Appender) appendRowArray(args []driver.Value) error {
 		}
 
 		if err := info.typeMatch(reflect.TypeOf(v)); err != nil {
-			return fmt.Errorf("type mismatch for column %d: \n%s \n%s", i, err.Error(), errNote)
+			return fmt.Errorf("type mismatch for column %d: \n%s", i, err.Error())
```
Comment (disq) [inline appender.go:551]:
```suggestion
			return fmt.Errorf("type mismatch for column %d: \n%s", i + 1, err.Error())
```

I think 1-indexed would be more clear since we're talking about columns and not slice elements in the error message.

**Comment by disq** [comment]
@taniabogatsch in `appendRowArray`, the untyped nil check (quoted below) doesn't seem to be enough for nullable list items.

```go
		if v == nil {
			setNull(&info, a.currSize)
			continue
		}
```

To be able to append a list with a null value in it, the slice has to be of type `[]*T`. But `typeMatch` won't allow it:

> type mismatch for column 29: expected: []int64, actual: []*int64

```
select column_name, ordinal_position, is_nullable, data_type from information_schema.columns where ordinal_position=30;
┌─────────────┬──────────────────┬─────────────┬───────────┐
│ column_name │ ordinal_position │ is_nullable │ data_type │
│   varchar   │      int32       │   varchar   │  varchar  │
├─────────────┼──────────────────┼─────────────┼───────────┤
│ id_list     │               30 │ YES         │ BIGINT[]  │
└─────────────┴──────────────────┴─────────────┴───────────┘
D 
```

File: appender.go (hunk 1 above)
Comment (taniabogatsch) [reply appender.go]: Fair point! I've updated both error messages (append row column type mismatch and table column type mismatch during appender creation).

File: appender.go [hunk 2]
```diff
@@ -236,169 +230,188 @@ func mallocLogicalTypeSlice(count int) (unsafe.Pointer, []C.duckdb_logical_type)
...
+}
 
-	csPtr := unsafe.Pointer(C.malloc(C.size_t(count) * size))
-	slice := (*[1 << 30]*C.char)(csPtr)[:count:count]
-	return csPtr, slice
+func getValue[T any](val any) T {
+	if reflect.ValueOf(val).Kind() != reflect.Ptr {
+		return val.(T)
+	}
+	return *(val.(*T))
```
Comment (disq) [inline appender.go:248]: There's a nil pointer dereference here, called from the `colInfo` generated in `initCString` (`blob` branch) and called from `setList`'s `childInfo.fn`

Comment (taniabogatsch) [reply appender.go:248]: Nice catch - I was missing an early out. On this note, we need to add more testing to these changes. We now allow passing pointers also for top-level types. I am already extending some test cases, but please drop anything here that you can find that still breaks!

**Comment by taniabogatsch** [comment]
Also, cc @marcboeker, what is your take on allowing (nil)-pointers as types in the appender? I understand it is an elegant solution to better support `NULL` types for now. Is there another way? We could potentially also move to the [`database/sql`](https://pkg.go.dev/database/sql#NullInt32) package `NULL` types. 🤔 

That is, we focus on column type initialization in this PR. Then, we can approach the whole `NULL` topic in a separate, dedicated PR. As I already noticed other tests breaking, if I extend them to this solution. cc @disq ?

File: appender.go (hunk 2 above)
Comment (disq) [reply appender.go:248]: Our test suite (of various types and lists of those types) is now passing without issue, thanks for the rapid responses.

**Comment by marcboeker** [comment]
#167 is merged now 🙂

**Comment by taniabogatsch** [comment]
@disq - fair point! I'll disentangle this PR to focus on the type creation/initialization. Then, I'll open a separate PR and port my `NULL` changes to that one. That way, we do not mix both, and I can focus on adding a few more tests for the `NULL`s. Eventually, we can discuss better `sql.Null*` type support in the appender/go-duckdb, as raised in https://github.com/marcboeker/go-duckdb/issues/89.
